File: pythagorean_triple_generator.clj
;; ピタゴラス数生成

(ns pythagorean-triple-generator
  (:require [clojure.math.numeric-tower :refer [expt gcd]]))

(defn pythagorean-triples []
  (for [c (iterate inc 1)
        b (range 1 (inc c))
        a (range 1 (inc b))
        :let [coprime (fn [x y z] (and (== (gcd x y) 1)
                                       (== (gcd y z) 1)))]
        :when (and (== (+ (expt a 2) (expt b 2)) (expt c 2))
                   (coprime a b c))]
    [a b c]))

;; 利用例
(println (take 10 (pythagorean-triples)))

File: PythagoreanTripleGenerator.hs
-- ピタゴラス数生成

module PythagoreanTripleGenerator (
  pythagoreanTriples,
  main
) where

pythagoreanTriples :: Integral a => [(a, a, a)]
pythagoreanTriples =
  [(a, b, c) | c <- [1..], b <- [1..c], a <- [1..b],
    a ^ 2 + b ^ 2 == c ^ 2, coprime a b c]
    where
      coprime x y z = gcd x y == 1 && gcd y z == 1

-- 利用例
main :: IO ()
main = print $ take 10 pythagoreanTriples
